src/main/host/futex.c

```diff
@@ -73,15 +75,29 @@ unsigned int futex_wake(Futex* futex, unsigned int numWakeups) {
     // in the status changed callback.
     GList* listenerList = g_hash_table_get_keys(futex->listeners);
 
-    // Iterate the listeners.
+    // Iterate the listeners and perform the requested number of wakeups if we can.
     GList* item = g_list_first(listenerList);
-    unsigned int numWoken;
-    for (numWoken = 0; item && numWoken < numWakeups; numWoken++) {
+    unsigned int numWoken = 0;
+
+    while (item && (numWoken < numWakeups)) {
         StatusListener* listener = item->data;
 
         // Only call if the listener is still valid
         if (g_hash_table_contains(futex->listeners, listener)) {
-            statuslistener_onStatusChanged(listener, STATUS_FUTEX_WAKEUP, STATUS_FUTEX_WAKEUP);
+            // If this listener was already woken up, skip it this time
+            bool did_wakeup =
```

Comment:
Would it make sense to add some `pending_wake` flag in the listener itself, instead of keeping an auxiliary table here?

I think that might be a behavior change if a task gets woken by multiple conditions before it runs again, but I'm not sure which behavior is right :)

Reply:
Hmm. The pending flag is already handled in the syscall condition:

https://github.com/shadow/shadow/blob/2d6ed8dc5840958aeda2ab264669f6f00a447e0c/src/main/host/syscall_condition.c#L336-L339

The problem is that the futex does not have visibility into the state of the syscall condition and so cannot query the pending flag. Maybe we could try to propagate the state of the pending flag back down the call stack, but we already use a table in futex anyway so storing the boolean is little extra complexity.

Reply:
There may be a cleaner way to do this, but I'm going to avoid that rabbit hole for now.